designs/0005-static-matrices.md

```diff
@@ -0,0 +1,147 @@
+- Feature Name: static_matrices
+- Start Date: April 30th, 2020
+- RFC PR: [Example Branch](https://github.com/stan-dev/math/compare/feature/var-template)
+- Stan Issue: [#1805](https://github.com/stan-dev/math/issues/1805)
+
+# Summary
+[summary]: #summary
+
+This proposes a constant matrix type in stan that will allow for significant performance opportunities. In Stan Math this will be represented by `var<Eigen::Matrix<double, R, C>>` or `var<matrix_cl<double>>`. The implementation will be fully backwards compatible with current Stan code. With optimizations from the compiler the conditions for a `matrix` to be a constant matrix can be detected and applied to Stan programs automatically for users. The implementation of this proposal suggests a staged approach where a stan language level feature is delayed until type attributes are allowed in the language and constant matrices have support for the same set of methods that the current `matrix` type has.
```

Comment:
`var<T>` -> `var_value<T>`

"With optimizations from the compiler the conditions for a `matrix` to be a constant matrix can be detected and applied to Stan programs automatically for users." -> what we've ended up working on is purely the math implementation stuff. I wouldn't mind if this design-doc shrunk to just the math stuff. At the language we'll either handle stuff automatically or introduce new types -- and likely both, but there will be some other solution.

Reply:
The only mentions of the compiler/language here are to say that this design doc is just focused on the math portion. The only other example is to show what things could be deduced as static by the compiler. Should I change that one just to show how to rewrite that particular model for static matrices?

Reply:
Lemme just open up a pull request to this with the changes I'm talking about and you can tell me if it's overboard. I think that'll be easier.